### PR TITLE
fix(es/modules): Fix re-assignment of the export variable in export default function

### PR DESCRIPTION
**Pull request by wre232114**
**Description:**

The swc_ecma_transforms_modules crate do not transform the following code as expected：

**Input**
```js
export var format;
export function defaultLocale(definition) {
  locale = formatLocale(definition);
  format = locale.format;
  return locale;
}
format = '123';
```

The expected generated code as below, which is the same as the output of babel:
**Expected Output**
```js
"use strict";
Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = defaultLocale;
exports.format = void 0;
function defaultLocale(definition) {
    locale = formatLocale(definition);
    exports.format = format = locale.format;
    return locale;
}
var format;
exports.format = format;
exports.format = format = '123';
```

the re-assignment of `format` in defaultLocale should be transformed to `exports.format = format = local.format`. current output:

**Current Unexpected Output**
```js
"use strict";
Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = defaultLocale;
exports.format = void 0;
function defaultLocale(definition) {
    locale = formatLocale(definition);
    format = locale.format;
    return locale;
}
var format;
exports.format = format;
exports.format = format = '123';
```

A test case is added for this case.

**BREAKING CHANGE:**

**Related issue (if exists):**

- Closes https://github.com/swc-project/swc/issues/4401

### CODE REVIEW

**Review by kdy1 — CHANGES_REQUESTED** [review]
Can you run `UPDATE=1 cargo test` in `./crates/swc`?
You may need to run it multiple time.

**Comment by wre232114** [comment]
@kdy1 updated

File: crates/swc_ecma_transforms_module/src/common_js.rs [hunk 1]
```diff
@@ -213,6 +213,18 @@ impl Fold for CommonJs {
                     for ident in found {
                         exports.push(ident.sym.clone());
                     }
+
+                    if let Decl::Var(VarDecl { decls, .. }) = decl {
+                        for var_decl in decls {
+                            if let VarDeclarator {
```
Comment (kdy1) [inline crates/swc_ecma_transforms_module/src/common_js.rs:219]: I think we should use `find_pat_ids` here instead.

Comment (kdy1) [reply crates/swc_ecma_transforms_module/src/common_js.rs:219]: Can you add a test case for exporting something with an array pattern or an object pattern?

File: crates/swc_ecma_transforms_module/src/util.rs [hunk 2]
```diff
@@ -161,6 +161,17 @@ pub struct Scope {
 }
 
 impl Scope {
+    /// insert exported bindings without duplication
+    pub(crate) fn insert_exported_bindings(&mut self, ident: &Ident) {
+        let bindings = self
+            .exported_bindings
+            .entry((ident.sym.clone(), ident.span.ctxt()))
+            .or_default();
+        if !bindings.contains(&(ident.sym.clone(), ident.span.ctxt())) {
```
Comment (kdy1) [inline crates/swc_ecma_transforms_module/src/util.rs:170]: Please use `ident.to_id()` instead

Comment (kdy1) [reply crates/swc_ecma_transforms_module/src/util.rs:170]: `(ident.sym.clone(), ident.span.ctxt())` is old version of `ident.to_id()`

File: crates/swc_ecma_transforms_module/src/util.rs [hunk 3]
```diff
@@ -161,6 +161,17 @@ pub struct Scope {
...
+            bindings.push((ident.sym.clone(), ident.span.ctxt()));
```
Comment (kdy1) [inline crates/swc_ecma_transforms_module/src/util.rs:171]: Same here

File: crates/swc_ecma_transforms_module/src/common_js.rs (hunk 1 above)
Comment (wre232114) [reply crates/swc_ecma_transforms_module/src/common_js.rs:219]: ok

File: crates/swc_ecma_transforms_module/src/util.rs (hunk 2 above)
Comment (wre232114) [reply crates/swc_ecma_transforms_module/src/util.rs:170]: ok

**Review by kdy1 — APPROVED** [review]
Thank you!

---

swc-bump:
 - swc_ecma_transforms_module

**Comment by wre232114** [comment]
The `Performance regression check` failed due to Network issue. Could you rerun it, thank you

**Pull request merged**